lib/extractor/image.ts

```diff
@@ -0,0 +1,145 @@
+    // (code has been modified to print the regex), and then adjusted to
+    // Javascript. The required modifications were replacing `[:xdigit:]` with
+    // `[a-fA-F0-9]` and escaping the slashes.
+    // Note that the digest matched in this Regex will match digests that have
+    // uppercase-letters, while the regex used in validateDigest does NOT match
+    // uppercase-letters. This simply matches the behaviour from the upstream
+    // `reference` and `go-digest `packages.
+    //
+    // we're matching pattern: <registry:port_number>(optional)/<image_name>(mandatory):<image_tag>(optional)@<tag_identifier>(optional)
+    // This Regex contains three capture groups:
+    // 1) The repository / image name
+    // 2) tag
+    // 3) digest
+    const re =
+      /^((?:(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+)?(?::[0-9]+)?\/)?[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:\/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?)(?::([\w][\w.-]{0,127}))?(?:@([A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][A-Fa-f0-9]{32,}))?$/gi;
+
+    const groups = re.exec(targetImage);
+    if (groups === null) {
+      if (targetImage === "") {
+        throw Error("image name is empty");
+      }
+      if (re.exec(targetImage.toLowerCase()) !== null) {
+        throw Error("image repository contains uppercase letter");
+      }
+      throw Error("invalid image reference format");
+    }
+
+    const GROUP = {
+      NAME: 1,
+      TAG: 2,
+      DIGEST: 3,
+    };
+
+    const NameTotalLengthMax = 255;
+    if (groups[GROUP.NAME].length > NameTotalLengthMax) {
+      throw Error("image repository name is more than 255 characters");
+    }
+    this.name = groups[GROUP.NAME];
+
+    this.tag =
+      !groups[GROUP.TAG] && !groups[GROUP.DIGEST]
+        ? "latest"
+        : groups[GROUP.TAG];
+
+    this.digests = {};
+    if (digests.index) {
+      this.digests[DigestType.Index] = new ImageDigest(digests.index);
+    }
+    if (digests.manifest) {
+      this.digests[DigestType.Manifest] = new ImageDigest(digests.manifest);
+    }
+
+    if (groups[GROUP.DIGEST]) {
+      const digest = new ImageDigest(groups[GROUP.DIGEST]);
+      if (
+        this.digests[DigestType.Manifest] !== digest &&
+        this.digests[DigestType.Index] !== digest
+      ) {
+        this.digests[DigestType.Unknown] = digest;
+      }
+    }
+  }
+
+  public getAllNames(): string[] {
+    const names: Set<string> = new Set();
+    if (this.tag) {
+      names.add(this.name + ":" + this.tag);
+    }
+    if (this.digests[DigestType.Manifest]) {
+      names.add(this.name + "@" + this.digests[DigestType.Manifest].toString());
+    }
+    if (this.digests[DigestType.Index]) {
+      names.add(this.name + "@" + this.digests[DigestType.Index].toString());
+    }
+    if (this.digests[DigestType.Unknown]) {
+      names.add(this.name + "@" + this.digests[DigestType.Unknown].toString());
+    }
```

Comment:
I think for JavaScript, I don't have this option, as you can't iterate an object, but its key or value, which they can't call the `toString()` method. 
https://www.freecodecamp.org/news/how-to-iterate-over-objects-in-javascript/

Reply:
I might totally be lying about can't iterate through an object, but I think the current format is easier to read, especially when the digest only has one entry, it might be overkill to use a loop? 😄

Reply:
> I think for JavaScript, I don't have this option, as you can't iterate an object, but its key or value, which they can't call the toString() method.

You are lying 😛 Afaik this should work:
```js
for (const digest of Object.values(this.digests)) {
  names.push(this.name+"@"+digest.toString())
}
```

I don't think it matters from a readability perspective to do `this.digests.unknown` vs `this.digests[DigestType.Unknown]`. 

Regarding the "to loop or not to loop": Also not sure I agree, but I also don't really mind 😄 

Keep it as is :)

Reply:
I know, that is what I said `I might totally be lying about can't iterate through an object` 😆. 

you can even use
```javascript
Object.values(this.digests).forEach(digest => names.push(this.name+"@"+digest.toString()));
```
However, I don't know whether writing it out specifically would help retain the order, as you can try this [test case](https://github.com/snyk/snyk-docker-plugin/pull/465/files#diff-22abcd98264342dbdf9b059a66cc7c55a1fb14c3bf9b65080b6870667699bfdbR27-R42), the order in the array was reversed if we use a loop. I don't know when iterating the object, whether the order would be preserved, or maybe we don't care about the order, but just need the array `includes()` these names?

I still prefer writing them out explicitly 🥺